8248268: Support KWP in addition to KW

valeriepeng opened the pull request:
This change updates SunJCE provider as below:
- updated existing AESWrap support with AES/KW/NoPadding cipher transformation. 
- added support for AES/KWP/NoPadding and AES/KW/PKCS5Padding.

Existing AESWrap impl, i.e. AESWrapCipher class, is re-factored and renamed to KeyWrapCipher class. The W and W_inverse functions are moved to KWUtil class. The KW and KWP support are in the new AESKeyWrap and AESKeyWrapPadded classes which extend FeedbackCipher and used in KeyWrapCipher class. To minimize data copying, AESKeyWrap and AESKeyWrapPadded will do the crypto operation over the same input buffer which is allocated and managed by KeyWrapCipher class. 

Also note that existing AESWrap impl does not take IV. However, the corresponding PKCS#11 mechanisms do, so I added support for accepting IVs to both KW and KWP.

Thanks,
Valerie

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8248268](https://bugs.openjdk.java.net/browse/JDK-8248268): Support KWP in addition to KW


### Reviewers
 * [Xue-Lei Andrew Fan](https://openjdk.java.net/census#xuelei) (@XueleiFan - **Reviewer**) ⚠️ Review applies to cfe66d54c554b2ddb24df32d2908326df89963a9


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2404/head:pull/2404` \
`$ git checkout pull/2404`

Update a local copy of the PR: \
`$ git checkout pull/2404` \
`$ git pull https://git.openjdk.java.net/jdk pull/2404/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2404`

View PR using the GUI difftool: \
`$ git pr show -t 2404`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2404.diff">https://git.openjdk.java.net/jdk/pull/2404.diff</a>

</details>

valeriepeng commented:
/csr

valeriepeng commented:
Ping, anyone has time to review this?

SalusaSecondus commented on src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java:
```diff
@@ -0,0 +1,753 @@
+            throws IllegalBlockSizeException, BadPaddingException {
+        if (!decrypting && dataIdx == 0) {
+            writeIvSemiBlock(dataBuf);
+        }
+
+        if (inLen > 0) {
+            store(in, inOfs, inLen);
+        }
+
+        // if enc, add padding
+        if (!decrypting) {
+            int actualLen = dataIdx - SEMI_BLKSIZE;
+            if (padding != null) {
+                int paddingLen = padding.padLength(actualLen);
+                // check and re-size dataBuf if needed
+                store(null, 0, paddingLen);
+                try {
+                    padding.padWithLen(dataBuf, dataIdx, paddingLen);
+                    dataIdx += paddingLen;
+                } catch (ShortBufferException sbe) {
+                    // should never happen
+                    throw new AssertionError();
+                }
+            }
+        }
+        try {
+            int outLen;
+            if (decrypting) {
+                outLen = cipher.decryptFinal(dataBuf, 0, dataIdx, null, -1);
+                // unpad if padding is used
+                if (padding != null) {
+                    int padIdx = padding.unpad(dataBuf, 0, outLen);
+                    if (padIdx <= 0) {
+                        throw new BadPaddingException("Bad Padding: " + padIdx);
+                    }
+                    outLen = padIdx;
+                }
+            } else {
+                outLen = cipher.encryptFinal(dataBuf, 0, dataIdx, null, -1);
+            }
```
Can we extract this shared logic (among the different versions of `engineDoFinal`) into a separate helper method so that the different `engineDoFinal` methods just need to handle their specific differences (such as returning `Arrays.copyOf(dataBuf, outLen)` or calling `System.arraycopy(dataBuf, 0, out, outOfs, outLen); return outLen;`).

SalusaSecondus commented on src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java:
```diff
@@ -0,0 +1,753 @@
+     * being passed to a software only cipher).
+     */
+    @Override
+    protected byte[] engineWrap(Key key)
+            throws IllegalBlockSizeException, InvalidKeyException {
```
Is it okay that we can call `cipher.init(Cipher.ENCRYPT_MODE, ...)` and then use it with `cipher.wrap()`?

Also, can we simply delegate the heavy lifting to `engineDoFinal()` (or the new suggested helper method) rather than duplicating this logic here?

valeriepeng replied:
Right, I will add additional checking on the Cipher.XXX_MODE. Agree with you regarding reducing the code duplication.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrap.java:
```diff
@@ -0,0 +1,196 @@
+ * "Recommendation for Block Cipher Modes of Operation: Methods for Key Wrapping"</a>
+ * and represents AES cipher in KW mode.
+ */
+class AESKeyWrap extends FeedbackCipher {
```
I chose to extend FeedbackCipher since we are supporting KW and KWP as cipher modes and given that FeedbackCipher represents "a block cipher in one of its modes", it seems a natural fit. As for the various unsupported operations, those methods aren't currently used and instead of a no-op, I chose to throw UnsupportedOperationException to prevent future accidental use.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrap.java:
```diff
@@ -0,0 +1,196 @@
+            throw new IllegalBlockSizeException("data should" +
+                " be at least 16 bytes and multiples of 8");
+        }
+        // assert ptOfs == 0; ct == pt; ctOfs == 0;
```
Hmm, there are some inconsistencies in how the variables are named. I will remove this comment and update the method javadoc accordingly, i.e. ptOfs, ct, ctOfs have all been renamed to dummyN.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrapPadded.java:
```diff
@@ -0,0 +1,254 @@
+    private static int validateIV(byte[] ivAndLen, byte[] icv)
+            throws IllegalBlockSizeException {
+        // check against icv and fail if not match
+        if (!Arrays.equals(ivAndLen, 0, ICV2.length, icv, 0, ICV2.length)) {
+            throw new IllegalBlockSizeException("Integrity check failed");
+        }
```
This is just IV and length, not plaintext. So, I didn't use the constant time array check. I can switch to the constant time version, it's trivial.

valeriepeng commented on test/jdk/com/sun/crypto/provider/Cipher/KeyWrap/TestGeneral.java:
```diff
@@ -0,0 +1,187 @@
+        ALGO = "AES/KWP/NoPadding";
+        System.out.println("Testing " + ALGO);
+        c = Cipher.getInstance(ALGO, "SunJCE");
+        for (int i = 0; i < MAX_KWP_PAD_LEN; i++) {
```
Sure, I can add some from the NIST zip which Michael StJohns mentioned to the NISTWrapKAT.java.

valeriepeng commented:
> 
> 
> _Mailing list message from [Michael StJohns](mailto:mstjohns@comcast.net) on [security-dev](mailto:security-dev@openjdk.java.net):_
> 
> On 3/23/2021 4:15 PM, Greg Rubin wrote:
> 
> > > 177:         System.out.println("Testing " + ALGO);
> > > 178:         c = Cipher.getInstance(ALGO, "SunJCE");
> > > 179:         for (int i = 0; i < MAX_KWP_PAD_LEN; i++) {
> > > I see that here (and earlier) we do test all padding lengths. I'd still like some KATs generated by a known good implementation to ensure that we are not just compatible with ourselves.
> 
> http://csrc.nist.gov/groups/STM/cavp/documents/mac/kwtestvectors.zip has
> the NIST test vectors.? See
> https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/mac/KWVS.pdf
> for details.
> 
> Mike
> 
> -------------- next part --------------
> An HTML attachment was scrubbed...
> URL: <https://mail.openjdk.java.net/pipermail/security-dev/attachments/20210323/e1a400db/attachment.htm>

Sure, I will add some, thanks Mike for the pointers.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrapPadded.java:
```diff
@@ -0,0 +1,254 @@
+
+        int outLen = validateIV(ivAndLen, this.iv);
+        // check padding bytes
+        int padLen = ctLen - outLen;
```
Sure, good idea.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrap.java:
```diff
@@ -0,0 +1,196 @@
+            throw new IllegalBlockSizeException
+                ("data should be at least 24 bytes and multiples of 8");
+        }
+        byte[] ivOut = new byte[ICV1.length];
+        W_INV(ct, ctLen, ivOut, embeddedCipher);
+        ctLen -= SEMI_BLKSIZE;
+
+        // check against icv and fail if not match
+        if (!Arrays.equals(ivOut, 0, ICV1.length, this.iv, 0, ICV1.length)) {
+            throw new IllegalBlockSizeException("Integrity check failed");
+        }
```
Hmm, another alternative is to move this logic to the underlying key wrap impl, i.e. AESKeyWrap and AESKeyWrapPadded, and just write 8-byte of 0s as place holder in KeyWrapCipher class.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/KWUtil.java:
```diff
@@ -0,0 +1,113 @@
+     * @param ivOut buffer for holding the recovered ICV semiblock
+     * @param cipher the initialized cipher object used
+     */
+    static final void W_INV(byte[] in, int inLen, byte[] ivOut,
+            SymmetricCipher cipher) {
```
Ok, I can update to make W() handles the IV semiblock overwrite.

valeriepeng commented on test/jdk/com/sun/crypto/provider/Cipher/KeyWrap/NISTWrapKAT.java:
```diff
@@ -35,58 +35,128 @@
+    }
+
+    @DataProvider
+    public Object[][] testData() {
```
Sure, there are some more from the NIST zip.

valeriepeng commented:
Hi Mike,

Please find comments in line below.

> KeyWrapCipher.java:
> 
> > 212      * Sets the padding mechanism of this cipher. Currently, only
> > 213      * "NoPadding" scheme is accepted for this cipher.
> > 214      *
> > 215      * @param padding the padding mechanism
> > 216      *
> > 217      * @exception NoSuchPaddingException if the requested padding mechanism
> > 218      * does not match the supported padding scheme
> > 219      */
> > 220     @Override
> > 221     protected void engineSetPadding(String padding)
> > 222             throws NoSuchPaddingException {
> > 223         if ((this.padding == null && !"NoPadding".equalsIgnoreCase(padding)) ||
> > 224                 this.padding instanceof PKCS5Padding &&
> > 225                 "PKCS5Padding".equalsIgnoreCase(padding)) {
> > 226             throw new NoSuchPaddingException();
> > 227         }
> > 228     }
> 
> Shouldn't this be rejecting PKCS5Padding?

Right, will fix this and update the javadoc above as well.
 
> Actually, I keep wondering why this isn't? AES/KW/NoPadding,
> AES/KW/KWPPadding and AES/KW/AutoPadding where the latter doesn't take a
> user provided IV, but figures out which of the two default IV values to
> use based on whether the input is a multiple of the blocksize or not?
> 
> Decrypting is similar - do the decryption, and check which IV you get to
> figure out padded or not padded.

The way I view it, both KW and KWP are modes. For interoperability, I referenced the corresponding mechanism definition from PKCS#11 v3.0 and base the java transformation on the corresponding PKCS#11 mechanisms. As for AES/KW/AutoPadding, it's an interesting idea, is there any other provider or library support it? To encrypt data of any size, we can just use AES/KWP/NoPadding. Is there additional benefit for this AutoPadding scheme?

Thanks,
Valerie
> 
> Mike
> 
> -------------- next part --------------
> An HTML attachment was scrubbed...
> URL: <https://mail.openjdk.java.net/pipermail/security-dev/attachments/20210323/0372393b/attachment-0001.htm>

SalusaSecondus reviewed (commented):
I'd advise against the AutoPadding scheme without more careful analysis and discussion. Have we seen either KW or KWP specifications which recommend that behavior?

My concern is that we've seen cases before where two different cryptographic algorithms could be selected transparently upon decryption and it lowers the security of the overall system. (A variant of in-band signalling.) The general consensus that I've been seeing in the (applied) cryptographic community is strongly away from in-band signalling and towards the decryptor fully specifying the algorithms and behavior prior to attempting decryption.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrapPadded.java:
```diff
@@ -0,0 +1,260 @@
+
+        for (int k = 5; k < SEMI_BLKSIZE; k++) {
+            if (outLen != 0) {
+                outLen <<= SEMI_BLKSIZE;
```
Yes, makes sense. I will update it.

valeriepeng commented on test/jdk/com/sun/crypto/provider/Cipher/KeyWrap/TestGeneral.java:
```diff
@@ -0,0 +1,187 @@
+            int maxPadLen) throws Exception {
+
+        System.out.println("input len: " + inLen);
+        c.init(Cipher.ENCRYPT_MODE, KEY, new IvParameterSpec(in, 0, ivLen));
```
In NISTWrapKAT.java, all of its test vectors do not use custom IV. So, the default IV scenario for KW and KWP are covered there. I can add another init(...) call which sets the IV based on the current IV to that test.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/AESKeyWrapPadded.java:
```diff
@@ -0,0 +1,260 @@
+    private static int validateIV(byte[] ivAndLen, byte[] iv)
+            throws IllegalBlockSizeException {
+        // check against iv and fail if not match
+        boolean match = true;
+        for (int i = 0; i < ICV2.length; i++) {
+            match &= (ivAndLen[i] == iv[i]);
+        }
+        if (!match) {
```
Sure, I can change to below:
```suggestion
        int match = 0;
        for (int i = 0; i < ICV2.length; i++) {
            match |= (ivAndLen[i] ^ iv[i]);
        }
        if (match != 0) {
            throw new IllegalBlockSizeException("Integrity check failed");
        }

```
Is this what you have in mind?

XueleiFan commented on src/java.base/share/classes/com/sun/crypto/provider/AESParameters.java:
```diff
@@ -46,7 +47,7 @@
     private BlockCipherParamsCore core;
 
     public AESParameters() {
-        core = new BlockCipherParamsCore(AESConstants.AES_BLOCK_SIZE);
+        core = new BlockCipherParamsCore(AESConstants.AES_BLOCK_SIZE, 4, 8);
```
A cipher object may not take different IV sizes at the same time.  I was just wondering how it could be used in practice.  Maybe something like:

```
AlgorithmParameters algParams = AlgorithmParameters.getInstance("AES");
algParams.init(ivParameterSpec);
```
The IV parameter is given with the init() method.  Then, it may be not necessary to construct the BlockCipherParamsCore object will all potential IV sized.  See the comments in BlockCipherParamsCore.

valeriepeng replied:
Cipher objects normally takes just one iv size. BlockCipherParamsCore is used by various impls of AlgorithmParametersSpi class which may be used with different block cipher algorithms. The iv parameter is given with the AlgorithmParametersSpi.init() method and invalid iv will lead to exceptions. Since there are iv size checks built in BlockCipherParamsCore already, it seems safer to relax the check a bit for AES (4 and 8 for KWP and KW). The other choice is to just remove the size check from BlockCipherParamsCore for AES algorithm, but then we'd not be able to reject invalid iv values as before.

XueleiFan commented on src/java.base/share/classes/com/sun/crypto/provider/BlockCipherParamsCore.java:
```diff
@@ -48,8 +49,11 @@
     private byte[] iv = null;
 
-    BlockCipherParamsCore(int blksize) {
+    private int[] moreSizes = null;
```
The moreSizes is not used other than the init() method field.  It may be not easy to check the specific size if we cache all supported sized in the object.  For example, if the required IV size if 8 bytes, it may be a problem about how to make sure the iv size is 8 bytes exactly for a specific algorithm.

Maybe, we could just have a ivSize filed.  The default value is block_size, which coupe be set with the init(ivParameterSpec) method.

```
    ....
    private int ivSize;
    ...
   BlockCipherParamsCore(int blkSize) {
       block_size = blkSize;
       ivSize = blkSize;
    }
    ...
   void init(AlgorithmParameterSpec paramSpec) {
        ivSize = ...;  // reset IV size.
    }

    // use ivSize while encoding and decoding.

```

valeriepeng replied:
The problem with this is that this assumes that the specified paramSpec argument of the init() method is always valid.

XueleiFan commented on src/java.base/share/classes/com/sun/crypto/provider/ConstructKeys.java:
```diff
@@ -161,29 +158,42 @@ private static final PrivateKey constructPrivateKey(byte[] encodedKey,
-                                                      keyAlgorithm);
-            break;
+            return ConstructKeys.constructPublicKey(encoding, keyAlgorithm);
+        default:
+            throw new RuntimeException("Unsupported key type");
```
Good improvement to thrown an exception rather than return "null"!

Is NoSuchAlgorithmException fitted better the specification?  See also the comment in  KeyWrapCipher.engineUnwrap().

valeriepeng replied:
Yes, will change.

XueleiFan commented:
Good points, Mike!  Thank you!

> _Mailing list message from [Michael StJohns](mailto:mstjohns@comcast.net) on [security-dev](mailto:security-dev@mail.openjdk.java.net):_
> > src/java.base/share/classes/com/sun/crypto/provider/AESParameters.java line 50:
> > > 48:
> > > 49:     public AESParameters() {
> > > 50:         core = new BlockCipherParamsCore(AESConstants.AES_BLOCK_SIZE, 4, 8);
> > > A cipher object may not take different IV sizes at the same time.  I was just wondering how it could be used in practice.  Maybe something like:
> 
> The mode is KW - it has a fixed length 8 byte non-iv integrity tag.???
> KWP is a special case of KW where there's still an 8 byte tag, but part
> of it is interpreted by KWP to figure out how much padding was
> included.?? KW (AKA RFC3394) permits user (actually specification
> specified) IV values.? KWP (aka RFC5649) does not.
> 
Hm, I missed this point.  KW (RFC 3394) supports Alternative IV (AIV), while KWP is just a case about how to construct the alternative IV for KWP operations.

> I'd treat KWP as a final (in the Java final sense) extension to KW with
> a fixed AIV flag value and a defined interpretation for the 8 byte AIV
> tag.? E.g. if you try to specify an IV for KWP, it should fail.?? If
> someone else wants to do something like KWP or even twiddle with the 4
> byte AIV, let them do their own KW wrap around - which they should be
> able to do that via the KW/NoPadding model by specifying their own AIV.?
> That should improve interoperability by preventing some monkey see
> monkey do errors.
> 

I agreed.  Maybe, we could do:
1. Support Alternative IV for KW, for the flexibility as described in section 2.2.3.2 of RFC 3394.
2. Use default IV if no AIV specified for KW, as described in section 2.2.3.1 of RFC 3394.
3. Use a fixed IV for KWP, as described in section 3of RFC 5649.

> This is sort of one reason I was arguing for AES/KW/KWPPadding rather
> than AES/KWP/NoPadding.
> 

I thought of the "AES/KW/KWPPadding" style as well.  The "AES/KWP/NoPadding" looks a little bit weird to me because there is padding while we call it with "NoPadding".

KWP is not exactly like the traditional AES/CBC/PKCS5Padding, where the components could be treated separately.  The padding scheme of KWP impact the IV as well.  So it was arguable to me.

Finally I feel better of the "AES/KWP/NoPadding" when I read the NIST SP title again, "Recommendation for Block Cipher Modes of Operation: Methods for Key Wrapping".  Maybe, it is an industry accepted notation to treat KWP as a block cipher mode.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/BlockCipherParamsCore.java:
```diff
@@ -59,9 +63,20 @@ void init(AlgorithmParameterSpec paramSpec)
+            throw new InvalidParameterSpecException("IV length not " +
+                        expectedLen + " bytes long");
         }
         iv = tmpIv.clone();
```
Right, good catch, I will fix it.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/ConstructKeys.java:
```diff
@@ -161,29 +158,42 @@ private static final PrivateKey constructPrivateKey(byte[] encodedKey,
+            if (ofs == 0 && len == encoding.length) {
+                return constructKey(encoding, keyAlgorithm, keyType);
+            } else {
+                byte[] encoding2 = Arrays.copyOfRange(encoding, ofs, ofs + len);
```
Right, maybe there is no constructor which takes offset and length due to trying to avoid the cost of validating them. I recall seeing other *Spec classes whose list of constructors are similar to X509EncodedKeySpec/PKCS8EncodedKeySpec as well.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/ConstructKeys.java:
```diff
@@ -161,29 +158,42 @@ private static final PrivateKey constructPrivateKey(byte[] encodedKey,
+                    keyAlgorithm);
+        } else {
+            if (ofs == 0 && len == encoding.length) {
+                return constructKey(encoding, keyAlgorithm, keyType);
+            } else {
+                byte[] encoding2 = Arrays.copyOfRange(encoding, ofs, ofs + len);
+                try {
+                    return constructKey(encoding2, keyAlgorithm, keyType);
```
Yes, I will refactor the two constructKey() methods as you suggested.

valeriepeng commented on src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java:
```diff
@@ -0,0 +1,722 @@
+        int outLen = 0;
+        try {
+            outLen = helperDecrypt(dataBuf, dataIdx);
+            return ConstructKeys.constructKey(dataBuf, 0, outLen,
+                    wrappedKeyAlgorithm, wrappedKeyType);
```
Sure, makes sense.

valeriepeng commented:
Current impl takes account into RFC 3394 (Sept 2002), RFC 5649 (Aug 2009), NIST 800-38F (Dec 2012) and PKCS#11 v3.0 Current Mech Spec (June 2020) and tries to support them all. If solely basing on RFC 3394/5649, then AES KW and KWP cipher should just supports key wrap/unwrap but not enc/dec. The main reason that I added the support for alternative iv and enc/dec is more for NIST 800-38F which states that KW/KWP can be used for general data protection as well and PKCS#11 v3.0 which lists 3 mechanisms mapping to java's AES using KW, AES using KW and PKCS5Padding and AES using KWP. These 3 mechanisms supports all 4, i.e. enc/dec/wrap/unwrap, as well custom IVs (8-byte for KW and 4-byte for KWP). Naming can be a very subjective matter and putting it aside, I don't see why KW and KWP should not allow custom IVs? Isn't hardcoded values less secure or fragile in general?

valeriepeng commented:
> 
> 
> _Mailing list message from [Michael StJohns](mailto:mstjohns@comcast.net) on [security-dev](mailto:security-dev@mail.openjdk.java.net):_
> 
> Some more general comments - related to the restructuring.
> 
> In AESKeyWrap at 152-155 - that check probably should be moved to W().??
> KWP should do the formatting prior to passing the data to W().? Also at
> 185-187 - move that to W_INV().
> 
> AESKeyWrap at 158 - shouldn't you be returning the cipher text length??
> That's what the comment in FeedbackCipher says.? W() should probably be
> returning the final length.
> 
> The length of the final ciphertext data should be 8 bytes longer than
> the plaintext. decryptFinal() seems to do the right thing by decreasing
> the length returned.?? But again - shouldn't that be the purview of W_INV()?
> 
> The call in KeyWrapCipher.engineGetOutputSize() should probably also be
> passed into KWUtil rather than being? done in KeyWrapCipher.? And the
> more I look at this, the more I think that all of the engineUpdate
> operations should throw UnsupportedOperationException - it would
> certainly simplify the logic.? That would make the call return either?
> inputLength + 8 or inputLength - 8 depending on mode.
> 
> KWUtil.W() - should probably check that in.length >= inLen + 8 and throw
> a ShortBufferException if not.
> 
> Would it be useful to add a comment in KeyWrapCipher that? warns
> maintainers that? AESKeyWrap(Padded).encryptFinal() and decryptFinal()
> uses the input buffer as the output buffer? That's a reasonable approach
> for decryption, but I'm wondering if you might want to support in-place
> encryption as there's no spec prohibition requiring data to be held
> until the encryption is complete.
> 
> Mike
> 
> On 5/24/2021 7:01 PM, Valerie Peng wrote:

Hi Mike,

Thanks for chiming in and review. Please find my comments below:
1) regarding the data length checks in AESKeyWrap/AESKeyWrapPadded class, it's better to check them before formatting the data and passing them to W(). This way, the error message can be clear and fail fast. I added assert statement to W()/W_INV() so the assumption on data sizes are clear.
2) regarding the returned text length of encryptFinal()/decryptFinal() calls, the returned value is correct in that 'ptLen' is the length of the formatted input to W() which contains both the first semiblock and user-supplied data. The name of the variable probably caused the confusion. For better readability, I changed W()/W_INV() to return an int. The javadoc explains what the returned int means. Hope this is clearer now.
3) regarding KWUtil, it's just a state-less class holding utility methods which are used by both AESKeyWrap (KW impl) and AESKeyWrapPadded (KWP impl). So I don't see why it should handle engineGetOutputSize() calls.
4) as for engineUpdate(...) throwing UnsupportedOperationException, I see the point that you are trying to make. Currently we are bound by the Cipher/CipherSpi javadoc spec. Another concern is that existing apps which are using multi-part enc/dec, may be caught off guard with this. So, that's how it is.

I need to do a system update and will get to your other comments once it's finished.
Thanks,
Valerie

valeriepeng commented:
In the latest commit, I have made some minor optimizations regrading the in-place encryption as suggested in Mike's earlier comments. However, the optimizations are limited to when the output offset == 0 case.

valeriepeng commented:
Hi Mike,
Lastly, regarding the naming, I am not too inclined toward the AES/KW/KWPPadding suggestion since the general concept of padding is that it's general, not mode-specific and are applied before data is processed by crypto operations. The mode can just process data without knowing whether padding is applied as long as the data has the right length. Thus, I'd keep KWP as a mode which does internal padding instead of a generic padding scheme. The AutoPadding suggestion is interesting and can be easily built on top of KW and KWP modes if desired. 

Thanks for the feedbacks,
Valerie

valeriepeng commented:
/integrate
